### PR TITLE
Add new CCCL:: CMake targets

### LINKED ISSUE #203: [FEA]: Update available cmake targets
### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

Infrastructure

### Is your feature request related to a problem? Please describe.

As a user of CCCL + cmake, I would like to have a convenient cmake target to link against that makes all of the CCCL code available on my include path as well as set any potential compiler flags.

### Describe the solution you'd like

I would like a `cccl::cccl` target that I can link against that provides Thrust/CUB/libcu++ to my application. 

```
target_link_libraries(my_app PRIVATE cccl::cccl)
```

This should effectively provide the same experience as if I were using CCCL packaged with the CTK, meaning I can `#include` anything from `<thrust/...>` `<cub/...>` `<cuda/...>`. Furthermore, the `thrust::device` backend should be `CUDA` and the `thrust::host` backend should be `cpp`. There is no expectation that using any of the other should be available. 

I would also like individual targets for Thrust/CUB/libcu++ exposed under the `cccl::` namespace, e.g., `cccl::thrust, cccl::cub, cccl::libcudacxx`. 

In order to optimize to make the common case easy and convenient, I'd like to not have to run any additional cmake functions to expose the targets other than `add_subdirectory` (or whatever else is used by things like CPM). 

### Describe alternatives you've considered

_No response_

### Additional context

For CUB and libcu++, I expect this will not be an issue, but this will be a departure from the status quo for Thrust where you currently have to invoke a cmake function in order to select which execution backends you want before the link targets are exposed. 

I believe the overwhelmingly common case is that people use the CUDA backend for `thrust::device` and sequential `cpp` backend for `thrust::host`. We should optimize to make this common case easy and convenient with the `cccl::cccl` target.

We can still require invoking a cmake function in order to expose more customized targets where people wish to configure alternative backends for `thrust::device/host`.

### PR DESCRIPTION
**Pull request by allisonvacanti**
## Description

closes #203

Add CCCL::CCCL, CCCL::libcudacxx, CCCL::CUB, and CCCL::Thrust targets.

The Thrust target is configured via cache options as discussed in #203.

## Checklist

- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by allisonvacanti** [comment]
@jrhemstad Needs tests and docs, but this works when I manually tested.

File: lib/cmake/cccl/cccl-config.cmake [hunk 1]
```diff
@@ -54,6 +70,25 @@ foreach(component IN LISTS components)
         "${cccl_cmake_dir}/../../../thrust/thrust/cmake/" # Source layout (GitHub)
         "${cccl_cmake_dir}/.."                            # Install layout
     )
+
+    if (TARGET Thrust::Thrust)
+      # By default, configure a CCCL::Thrust target with host=cpp device=cuda
+      option(CCCL_ENABLE_DEFAULT_THRUST_TARGET
```
[the remark this answers is not in the document]
Comment (allisonvacanti) [reply lib/cmake/cccl/cccl-config.cmake:76]: 👍

File: lib/cmake/cccl/cccl-config.cmake [hunk 2]
```diff
@@ -36,6 +40,12 @@ foreach(component IN LISTS components)
         "${cccl_cmake_dir}/../../../libcudacxx/lib/cmake/" # Source layout (GitHub)
         "${cccl_cmake_dir}/.."                             # Install layout
     )
+    # Can't alias other alias targets, so use the uglified target name instead
+    # of libcudacxx::libcudacxx:
+    if (TARGET _libcudacxx_libcudacxx)
+      add_library(CCCL::libcudacxx ALIAS _libcudacxx_libcudacxx)
```
Comment (robertmaynard) [inline lib/cmake/cccl/cccl-config.cmake:46]: Need to guard this on the CCCL::libcudacxx target already existing

File: lib/cmake/cccl/cccl-config.cmake [hunk 3]
```diff
@@ -45,6 +55,12 @@ foreach(component IN LISTS components)
         "${cccl_cmake_dir}/../../../cub/cub/cmake/" # Source layout (GitHub)
         "${cccl_cmake_dir}/.."                      # Install layout
     )
+    # Can't alias other alias targets, so use the uglified target name instead
+    # of CUB::CUB:
+    if (TARGET _CUB_CUB)
```
Comment (robertmaynard) [inline lib/cmake/cccl/cccl-config.cmake:60]:
```suggestion
    if (TARGET _CUB_CUB AND NOT TARGET CCCL::CUB)
```

File: lib/cmake/cccl/cccl-config.cmake [hunk 4]
```diff
@@ -54,6 +70,27 @@ foreach(component IN LISTS components)
         "${cccl_cmake_dir}/../../../thrust/thrust/cmake/" # Source layout (GitHub)
         "${cccl_cmake_dir}/.."                            # Install layout
     )
+
+    if (TARGET Thrust::Thrust)
```
Comment (robertmaynard) [inline lib/cmake/cccl/cccl-config.cmake:74]:
```suggestion
    if (TARGET Thrust::Thrust AND NOT TARGET CCCL::Thrust)
```

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: lib/cmake/cccl/cccl-config.cmake (hunk 2 above)
Comment (allisonvacanti) [reply lib/cmake/cccl/cccl-config.cmake:46]: Good catch -- fixing.

**Review by robertmaynard — APPROVED** [review]

**Comment by jrhemstad** [comment]
@allisonvacanti I think you still need to cherry-pick the commits you made to https://github.com/NVIDIA/cccl/pull/177 to this PR too.

**Comment by jrhemstad** [comment]
The libcu++ failure is a sporadic failure that can be ignored and I'm working on resolving the root cause.

**Comment by allisonvacanti** [comment]
> cherry-pick the commits

🤦🏻‍♀️ 

On it.

**Comment by allisonvacanti** [comment]
Should be all set now. I'll come back and check CI but feel free to merge sooner if needed.

**Review by jrhemstad — APPROVED** [review]

**Pull request merged**